File: backend/app/orders/order_service.py
from app.auth.auth_service import auth_service


class OrderService:
    """
    Service for retrieving and placing orders.
    """

    def get_order_book(self):
        response = auth_service.client.get_order_book()
        return response.json()

    def get_trade_book(self):
        response = auth_service.client.get_trade_book()
        return response.json()

    def place_order(
        self,
        variety,
        tradingsymbol,
        exchange,
        transaction_type,
        order_type,
        quantity,
        product,
        validity,
        price,
        trigger_price,
        disclosed_quantity,
        tag,
    ):
        try:
            response = auth_service.client.place_order(
                variety,
                tradingsymbol,
                exchange,
                transaction_type,
                order_type,
                quantity,
                product,
                validity,
                price,
                trigger_price,
                disclosed_quantity,
                tag,
            )

            return response.json()

        except Exception as e:
            return {
                "success": False,
                "error": str(e),
                "type": type(e).__name__,
            }


order_service = OrderService()

File: frontend/src/services/orderService.ts
const API_URL = "http://127.0.0.1:8000";

export type OrderRequest = {
  tradingsymbol: string;
  transaction_type: "BUY" | "SELL";
  quantity: number;
  order_type: "MARKET" | "LIMIT";
  price?: number;
};

export async function getOrders() {
  const response = await fetch(`${API_URL}/orders`);

  if (!response.ok) {
    throw new Error("Failed to fetch orders");
  }

  const data = await response.json();

  console.log("ORDER RESPONSE:", data);

  return data;
}

export async function placeOrder(order: OrderRequest) {
  console.log("ORDER OBJECT:", order);

  const params = new URLSearchParams({
    variety: "REGULAR",
    tradingsymbol: order.tradingsymbol,
    exchange: "NSE",
    transaction_type: order.transaction_type,
    order_type: order.order_type,
    quantity: order.quantity.toString(),
    product: "CNC",
    validity: "DAY",
    price: (order.price ?? 0).toString(),
    trigger_price: "0",
    disclosed_quantity: "0",
    tag: "FALCON",
  });

  console.log("REQUEST:", `${API_URL}/orders/place?${params.toString()}`);

  const response = await fetch(`${API_URL}/orders/place?${params.toString()}`, {
    method: "POST",
  });

  if (!response.ok) {
    throw new Error("Failed to place order");
  }

  return response.json();
}